api/genericresource/helpers.go

```diff
@@ -0,0 +1,122 @@
+package genericresource
+
+import (
+	"fmt"
+
+	"github.com/docker/swarmkit/api"
+)
+
+func newError(format string, args ...interface{}) error {
+	return fmt.Errorf("could not parse GenericResource: "+format, args...)
+}
+
+// NewSet creates a set object
+func NewSet(key string, vals ...string) []*api.GenericResource {
+	rs := make([]*api.GenericResource, 0, len(vals))
+
+	for _, v := range vals {
+		rs = append(rs, NewString(key, v))
+	}
+
+	return rs
+}
+
+// NewString creates a String resource
+func NewString(key, val string) *api.GenericResource {
+	return &api.GenericResource{
+		Resource: &api.GenericResource_Str{
+			Str: &api.GenericString{
+				Kind:  key,
+				Value: val,
+			},
+		},
+	}
+}
+
+// NewDiscrete creates a Discrete resource
+func NewDiscrete(key string, val int64) *api.GenericResource {
+	return &api.GenericResource{
+		Resource: &api.GenericResource_Discrete{
+			Discrete: &api.GenericDiscrete{
+				Kind:  key,
+				Value: val,
+			},
+		},
+	}
+}
+
+// NewResources creates a Resources object
+func NewResources() *api.Resources {
+	return &api.Resources{
+		Generic: make([]*api.GenericResource, 0),
+	}
+}
+
+func getKind(res *api.GenericResource) string {
+	switch r := res.Resource.(type) {
+	case *api.GenericResource_Discrete:
+		return r.Discrete.Kind
+	case *api.GenericResource_Str:
+		return r.Str.Kind
+	}
+
+	return ""
+}
+
+// GetResource returns resources from the "resources" parameter matching the kind key
+func GetResource(kind string, resources []*api.GenericResource) (res []*api.GenericResource) {
+	res = make([]*api.GenericResource, 0)
+	for _, r := range resources {
+		if getKind(r) != kind {
+			continue
+		}
+
+		res = append(res, r)
+	}
+
+	return res
+}
+
+// RemoveNodeResources removes "res" from nodeAvailable
+func RemoveNodeResources(nodeAvailable *[]*api.GenericResource, res []*api.GenericResource) {
+	if nodeAvailable == nil {
+		return
+	}
+
+	w := 0
+
+loop:
+	for _, na := range *nodeAvailable {
+		for _, r := range res {
+			if getKind(na) != getKind(r) {
+				continue
+			}
+
+			switch tr := r.Resource.(type) {
+			case *api.GenericResource_Discrete:
+				if na.GetDiscrete() == nil {
+					break // Type change, ignore
+				}
+
+				na.GetDiscrete().Value -= tr.Discrete.Value
+				if na.GetDiscrete().Value == 0 {
+					continue loop
+				}
+			case *api.GenericResource_Str:
+				if na.GetStr() == nil {
+					break // Type change, ignore
+				}
+
+				if tr.Str.Value != na.GetStr().Value {
+					break // not the right item, ignore
+				}
+				continue loop
```

Comment:
This is because if it is an item that needs to be removed then it shouldn't be written to `nodeAvailable`.
It is tested at a higher level in resource_management_test by testing the Claim function.

Do you want me to unit test it ?